### PR TITLE
Batch updates to global heapSize

### PR DESCRIPTION
**Pull request by adamlerer**
Contribution from @colesbury

Multithreaded Torch programs that do many small THAllocs were
encountering contention on updating heapSize. Instead, keep track of
allocations and frees in a thread-local variable, heapDelta, and only
update the global heapSize when heapDelta exceeds +/- 1MB.

### CODE REVIEW

**Comment by soumith** [comment]
With this PR, is https://github.com/torch/torch7/pull/353 still relevant?

**Comment by adamlerer** [comment]
nope, I abandoned it

**Comment by soumith** [comment]
I'll merge this in tomorrow unless there are objections. cc: @dominikgrewe , we hit a pretty serious perf-regression without this patch, when we have 30 threads doing hogwild.

File: lib/TH/THGeneral.c [hunk 1]
```diff
@@ -139,10 +141,14 @@ static long getAllocSize(void *ptr) {
...
 static void maybeTriggerGC(long curHeapSize) {
-  if(torchGCFunction && curHeapSize > heapSoftmax ) {
+  if (torchGCFunction && curHeapSize > heapSoftmax) {
     torchGCFunction(torchGCData);
-    long newHeapSize = THAtomicGetLong(&heapSize);
-    if(newHeapSize > heapSoftmax * heapSoftmaxGrowthThresh) {
+
+    // ensure heapSize is accurate before updating heapSoftmax
+    long newHeapSize = THAtomicAddLong(&heapSize, heapDelta) + heapDelta;
+    heapDelta = 0;
```
Comment (dominikgrewe) [inline lib/TH/THGeneral.c]: nit: Maybe factor this out into its own function, something like `applyHeapDelta` or `updateHeapSize`? I know it's only two lines of code but it's easy to get wrong, e.g. forgetting to reset heapDelta.

**Comment by dominikgrewe** [comment]
Looks good to me. I guess we should do the same in cutorch then.

**Comment by dominikgrewe** [comment]
Just saw the cutorch version :)

**Pull request merged**